### DISCUSSION TITLE
How to append response headers to ServeFile?

### DISCUSSION THREAD

**Discussion by wbenny** [tokio-rs/axum#1686] [Q&A] [answered]
I went through various questions here, I spent hours in axum and tower code, and all I can conclude is that I can't simply modify response headers with `ServeFile` (e.g. add `Content-Disposition`, like with actix `NamedFile::set_content_disposition`).

Please tell me that I'm wrong and that I'm overlooking something, because I refuse to believe such fundamental functionality wouldn't be included.

Please, don't tell me to open the file manually and convert it to `StreamBody`, because that loses all the goodies `ServeFile` offers (like handling of range headers).

There's a similar discussion here: https://github.com/tokio-rs/axum/discussions/608 however that discussion ends exactly at the point where I'm picking this up.

I'm becoming quite hopeless. Is there a way how to do it, without reimplementing the whole `ServeFile`?

**Comment by davidpdrsn** [comment, marked as answer]
You can use https://docs.rs/tower-http/0.3.5/tower_http/set_header/response/index.html or https://docs.rs/tower/0.4.13/tower/trait.ServiceExt.html#method.map_response to add additional headers to a response.

**Reply by wbenny** [reply]
Thank you!

**Reply by wbenny** [reply]
Would it be possible to provide a simple example, please?

```rust
    let download = routing::get_service(
        ServeFile::new(www_path.join("test.mkv"))
    ).handle_error(handle_error);
    
    // this doesn't appear to work
    let download = download.map_response(
        |record|
        ([(header::CONTENT_DISPOSITION, "attachment; filename=override.mkv")], record)
            .into_response()
    );

```

**Reply by davidpdrsn** [reply]
`record` is really more `response` so you can do `response.headers_mut().insert(key, value)`. See the docs for [`http::Response`](https://docs.rs/http/0.2.8/http/response/struct.Response.html)

**Reply by wbenny** [reply]
This doesn't work either :/

```rust
    let download = download.map_response(
        |record| {
        record
            .headers_mut()
            .insert(
                header::CONTENT_DISPOSITION,
                HeaderValue::from_static("attachment; filename=override.mkv")
            );
        record
    });
```

**Reply by davidpdrsn** [reply]
Please elaborate. "Doesn't work" how?

**Reply by wbenny** [reply]
It doesn't compile. Usage:

```rust
    let app = Router::new()
        .route("/download", download);
/* mismatched types expected struct `MethodRouter<Arc<AppState>, _>`
   found struct `tower::util::MapResponse<MethodRouter<(), _> */
```

**Reply by davidpdrsn** [reply]
Use `.route_service(...)` instead of `.route(...)`

**Reply by wbenny** [reply]
Thanks a lot! Got it working now :)

**Comment by kreibaum** [comment]
Thanks for posting this question and answer, it was a great starting point.

The answer from 2023-01-08 not work for me anymore on axum 0.7.2. I ran into Send/Sync issues.
I found a different solution that works 2024-05-12.

Here is the static file where I need to add a header:

```rust
.nest_service(
    "/js/lib_worker.min.js",
    ServeFile::new("../target/js/lib_worker.min.js").precompressed_br(),
)
```

I first define a function to add the header to the response:

```rust
async fn credentialless<B>(mut response: Response<B>) -> Response<B> {
    response.headers_mut().insert("cross-origin-embedder-policy", "credentialless".parse().unwrap());
    response
}
```

Now I can do a similar trick as above, first putting the download service in a variable and then layering a `map_response` on top of it to add the header:

See: https://docs.rs/axum/latest/axum/middleware/fn.map_response.html

```rust
let download: routing::MethodRouter = routing::get_service(
    ServeFile::new("../target/js/lib_worker.min.js").precompressed_br()
).layer(map_response(credentialless));
```

Finally, I can nest the download service in the router:

```rust
.route_service(
    "/js/lib_worker.min.js",
    download,
)
```

If anyone is curious about the "off topic" bits in the example:

* `precompressed_br()` - See https://docs.rs/tower-http/latest/tower_http/services/struct.ServeFile.html#method.precompressed_br
* `"cross-origin-embedder-policy"` - Required for Browsers to activate some performance features.
